File: 8-String-to-Integer/string-to-integer.cpp
#include <bits/stdc++.h>
using namespace std;

class Solution {
public:
    int myAtoi(string s) {
        int f=1, i=0, n=s.size();
        long long num=0;
        while(i<n && s[i]==' '){
            i++;
        }
        if(s[i] == '-'){
            f = f*-1;
            i++;
        }
        else if(s[i] == '+'){
            i++;
        }
        while(isdigit(s[i])){
            num = num*10 + (s[i]-'0');
            if(num>INT_MAX && f==-1)
                return INT_MIN;
            else if(num>INT_MAX && f==1)
                return INT_MAX;
            i++;
        }
        return num*f;
    }
};

File: 8-String-to-Integer/string-to-integer.py
class Solution:
    def myAtoi(self, s: str) -> int:
        i, n = 0, len(s)
        f = 1
        num = 0

        while i < n and s[i] == ' ':
            i += 1

        if i < n and s[i] == '-':
            f = -1
            i += 1
        elif i < n and s[i] == '+':
            i += 1

        while i < n and s[i].isdigit():
            num = num * 10 + int(s[i])
            if num > 2**31 - 1 and f == 1:
                return 2**31 - 1
            elif num > 2**31 and f == -1:
                return -2**31
            i += 1

        return f * num
